src/dune_pkg/opam.ml

```diff
@@ -0,0 +1,188 @@
+
+  (* Version to use for local packages with no version number *)
+  let local_package_default_version = OpamPackage.Version.of_string "LOCAL"
+
+  type t =
+    { dir_context : Dir_context.t
+    ; local_packages : OpamFile.OPAM.t OpamPackage.Name.Map.t
+    }
+
+  type rejection = Dir_context.rejection
+
+  let pp_rejection = Dir_context.pp_rejection
+
+  let candidates t name =
+    match OpamPackage.Name.Map.find_opt name t.local_packages with
+    | None -> Dir_context.candidates t.dir_context name
+    | Some opam_file ->
+      let version =
+        Option.value opam_file.version ~default:local_package_default_version
+      in
+      [ (version, Ok opam_file) ]
+
+  let user_restrictions t = Dir_context.user_restrictions t.dir_context
+
+  let filter_deps t = Dir_context.filter_deps t.dir_context
+
+  let create ~env ~repo ~local_packages =
+    let env name = Env.find_by_name env ~name in
+    let { Repo.packages_dir_path } = repo in
+    let dir_context =
+      Dir_context.create ~prefer_oldest:true
+        ~constraints:OpamPackage.Name.Map.empty ~env packages_dir_path
+    in
+    { dir_context; local_packages }
+end
+
+module Solver = Opam_0install.Solver.Make (Solver_context)
+
+module Summary = struct
+  type t = { opam_packages_to_lock : OpamPackage.t list }
+
+  let selected_packages_message t =
+    User_message.make
+      (Pp.tag User_message.Style.Success
+         (Pp.text "Selected the following packages:")
+      :: List.map t.opam_packages_to_lock ~f:(fun package ->
+             Pp.textf "%s" (OpamPackage.to_string package)))
+end
+
+let opam_package_to_lock_file_pkg ~repo ~local_packages ~lock_dir_path
+    opam_package =
+  let name = OpamPackage.name opam_package in
+  let version =
+    OpamPackage.version opam_package |> OpamPackage.Version.to_string
+  in
+  let dev = OpamPackage.Name.Map.mem name local_packages in
+  let info =
+    { Lock_dir.Pkg_info.name =
+        Package_name.of_string (OpamPackage.Name.to_string name)
+    ; version
+    ; dev
+    ; source = None
+    }
+  in
+  let opam_file =
+    match OpamPackage.Name.Map.find_opt name local_packages with
+    | None -> Repo.load_opam_package repo opam_package
+    | Some local_package -> local_package
+  in
+  (* This will collect all the atoms from the package's dependency formula regardless of conditions *)
+  let deps =
+    OpamFormula.fold_right
+      (fun acc (name, _condition) -> name :: acc)
+      [] opam_file.depends
+    |> List.map ~f:(fun name ->
+           Package_name.of_string (OpamPackage.Name.to_string name))
+  in
+  { Lock_dir.Pkg.build_command = None
+  ; install_command = None
+  ; deps
+  ; info
+  ; lock_dir = lock_dir_path
+  ; exported_env = []
+  }
+
+let solve_package_list local_packages ~env ~repo =
+  let context = Solver_context.create ~env ~repo ~local_packages in
+  let result =
+    try Solver.solve context (OpamPackage.Name.Map.keys local_packages)
+    with exn ->
```

Comment:
Is this reachable at all? It seems weird that this function both raises and returns `result`.

Reply:
`Solver.solve` returns `Error` when it can't find a solution to the dependencies but it can also raise exceptions, for example if opam is unable to parse an opam file in the package repo. I'll add a comment to this effect to clarify why this is needed.

Reply:
Seems like a weird choice for an API to me.

In any case, if some exceptions are expected we should list them here and have proper dedicated messages for them. Doing `Exn.to_dyn exn |> Dyn.to_string` is going to show something rather ugly to users.

Some tests for these exceptions would be quite handy as well.

Reply:
Ok cool, I'll add a list of exceptions. Would it be reasonable to keep the `Exn.to_dyn exn |> Dyn.to_string` for exceptions not in the list though, as the alternative is that it will crash dune in this case? It will be hard to be sure we're handling all possible exceptions from opam, and we might update our opam port one day and find that it added a new exception without us noticing.

Reply:
I think it would be reasonable to ask opam (or the solver) to allow us to get a readable user message from any exception that might be user facing. 

> and we might update our opam port one day and find that it added a new exception without us noticing

That would mean we have a poor error message for an error state that is reachable by the user. Which is a bug our end and a `Code_error` would be appropriate.

Reply:
Cool. For this PR I'll catch all the exceptions from opam that I'm aware of and produce a readable `User_error` for each, and then have a catch all that produces a `Code_error` indicating that there's a user reachable error state that we haven't accounted for.